### ISSUE TITLE
rpm_package silently fails on packages with a `~` character in the version

### ISSUE THREAD

**Issue by irvingpop** [chef/chef#3496]
Recent packages created by Chef CI have a tilde character in the version, like so:  `supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm`  - and installation of these is silently failing on Chef 12.3.0
(note: I'm not 100% sure it is the tilde, but this is my best guess)

Given the following cookbook code:

```ruby
case node['platform_family']
when 'rhel'
  package = 'https://web-dl.packagecloud.io/chef/stable/packages/el/5/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm'
when 'debian'
  package = 'https://web-dl.packagecloud.io/chef/stable/packages/ubuntu/precise/supermarket_1.10.1~alpha.0-1_amd64.deb'
end

pkgname = ::File.basename(package)
cache_path = ::File.join(Chef::Config[:file_cache_path], pkgname)

remote_file cache_path do
  source package
  mode '0644'
end

if node['platform_family'] == 'rhel'
  rpm_package cache_path
  execute 'rpm -qi supermarket'
elsif node['platform_family'] == 'debian'
  dpkg_package cache_path
  execute 'dpkg -L supermarket'
end
```

It works correctly on Ubuntu, but on RHEL (tested on CentOS 5.11 and 6.6) the following happens:
```
           * rpm_package[supermarket] action install[2015-06-05T23:20:53+00:00] INFO: Processing rpm_package[supermarket] action install (/tmp/kitchen/cache/cookbooks/chef-server-ingredient/libraries/chef_server_ingredients_provider.rb line 62)
       [2015-06-05T23:20:53+00:00] DEBUG: providers for generic rpm_package resource enabled on node include: [Chef::Provider::Package::Rpm]
       [2015-06-05T23:20:53+00:00] DEBUG: providers that refused resource rpm_package[supermarket] were: []
       [2015-06-05T23:20:53+00:00] DEBUG: providers that support resource rpm_package[supermarket] include: [Chef::Provider::Package::Rpm]
       [2015-06-05T23:20:53+00:00] DEBUG: providers that survived replacement include: [Chef::Provider::Package::Rpm]
       [2015-06-05T23:20:53+00:00] DEBUG: rpm_package[supermarket] checking rpm status
       supermarket 1.10.1~alpha.0-1.el5
       [2015-06-05T23:20:53+00:00] DEBUG: rpm_package[supermarket] checking install state
       package supermarket is not installed
       [2015-06-05T23:20:53+00:00] DEBUG: rpm_package[supermarket] supermarket not installed, installing 
       [2015-06-05T23:20:53+00:00] DEBUG: rpm_package[supermarket] is already installed - nothing to do
        (up to date)
            (up to date)
```

I can install the package manually and that works just fine:

```
[root@default-centos-511 cache]# rpm -ivh supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm 
Preparing...                ########################################### [100%]
   1:supermarket            ########################################### [100%]

Thank you for installing Supermarket!

The next step in the process is to run:

supermarket-ctl reconfigure

[root@default-centos-511 cache]# 
[root@default-centos-511 cache]# rpm -qi supermarket
Name        : supermarket                  Relocations: / 
Version     : 1.10.1~alpha.0                    Vendor: Omnibus <omnibus@getchef.com>
Release     : 1.el5                         Build Date: Mon 18 May 2015 08:04:01 PM UTC
Install Date: Fri 05 Jun 2015 09:05:47 PM UTC      Build Host: ip-10-194-12-70.us-west-2.compute.internal
Group       : default                       Source RPM: supermarket-1.10.1~alpha.0-1.el5.src.rpm
Size        : 1141070737                       License: unknown
Signature   : (none)
Packager    : Chef Supermarket Team <supermarket@getchef.com>
URL         : https://supermarket.getchef.com
Summary     : The full stack of supermarket
Description :
The full stack of supermarket
```

**Comment by irvingpop** [comment]
FWIW:  I tested this on a few earlier versions of Chef.   Here's what I found
Also silently fails on:  12.2.1, 12.1.2
12.0.3 and 11.18.2 explicitly error out like:
```

           ================================================================================
           Error executing action `install` on resource 'rpm_package[/tmp/kitchen/cache/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm]'
           ================================================================================
           
           Chef::Exceptions::Package
           -------------------------
           No version specified, and no candidate version available for /tmp/kitchen/cache/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/package_test/recipes/default.rb
           
            18:   rpm_package cache_path
            19:   execute 'rpm -qi supermarket'
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/package_test/recipes/default.rb:18:in `from_file'
           
           rpm_package("/tmp/kitchen/cache/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm") do
             provider Chef::Provider::Package::Rpm
             action :install
             retries 0
             retry_delay 2
             guard_interpreter :default
             package_name "/tmp/kitchen/cache/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm"
             source "/tmp/kitchen/cache/supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm"
             timeout 900
             cookbook_name "package_test"
             recipe_name "default"
           end
           
       

```

**Comment by irvingpop** [comment]
I think the tl;dr here is twofold:
1.  Chef client should probably be more clear about not liking the package version like old versions (or else accept it)   cc:  @chef/client-core 
2.  Our CI should stop producing packages with tilde in the version  cc: @chef/release-engineers

**Comment by irvingpop** [comment]
It looks like this commit to omnibus by @schisamo is where the package naming change came from: https://github.com/chef/omnibus/commit/0d5e3d137719034e4b21d9b75302843702b575a0

I understand the reasoning, but sadly RHEL6 ships with rpm 4.8.    I also tested on CentOS 7.1 (rpm 4.11.1) but it also silently fails in the same way.

**Comment by schisamo** [comment]
@irvingpop This is most likely a bug in the provider only as the test stage of our CI pipelines have been installing RPMs with tildes in the name (using raw RPM commands) for months now. For example:
http://wilson.ci.chef.co/view/Supermarket/job/supermarket-test/architecture=x86_64,platform=el-6,project=supermarket,role=tester/

We use __RHEL 5__, __6__, and __7__ in our CI pipelines.

I was also able to install the package references in your ticket in CentOS 5, 6 and 7:

### CentOS 5
```
[vagrant@slave-centos-511 ~]$ cat /etc/redhat-release
CentOS release 5.11 (Final)
[vagrant@slave-centos-511 ~]$ rpm --version
RPM version 4.4.2.3
[vagrant@slave-centos-511 ~]$ sudo rpm -Uvh supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm
Preparing...                ########################################### [100%]
   1:supermarket            ########################################### [100%]

Thank you for installing Supermarket!

The next step in the process is to run:

supermarket-ctl reconfigure

[vagrant@slave-centos-511 ~]$ rpm -qi supermarket
Name        : supermarket                  Relocations: /
Version     : 1.10.1~alpha.0                    Vendor: Omnibus <omnibus@getchef.com>
Release     : 1.el5                         Build Date: Mon 18 May 2015 08:04:01 PM UTC
Install Date: Sat 06 Jun 2015 08:46:36 PM UTC      Build Host: ip-10-194-12-70.us-west-2.compute.internal
Group       : default                       Source RPM: supermarket-1.10.1~alpha.0-1.el5.src.rpm
Size        : 1141070737                       License: unknown
Signature   : (none)
Packager    : Chef Supermarket Team <supermarket@getchef.com>
URL         : https://supermarket.getchef.com
Summary     : The full stack of supermarket
Description :
The full stack of supermarket
```

### CentOS 6
```
[vagrant@slave-centos-66 ~]$ cat /etc/redhat-release
CentOS release 6.6 (Final)
[vagrant@slave-centos-66 ~]$ rpm --version
RPM version 4.8.0
[vagrant@slave-centos-66 ~]$ sudo rpm -Uvh supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm
Preparing...                ########################################### [100%]
   1:supermarket            ########################################### [100%]

Thank you for installing Supermarket!

The next step in the process is to run:

supermarket-ctl reconfigure

[vagrant@slave-centos-66 ~]$ rpm -qi supermarket
Name        : supermarket                  Relocations: /
Version     : 1.10.1~alpha.0                    Vendor: Omnibus <omnibus@getchef.com>
Release     : 1.el5                         Build Date: Mon 18 May 2015 08:04:01 PM UTC
Install Date: Sat 06 Jun 2015 08:44:33 PM UTC      Build Host: ip-10-194-12-70.us-west-2.compute.internal
Group       : default                       Source RPM: supermarket-1.10.1~alpha.0-1.el5.src.rpm
Size        : 1141070737                       License: unknown
Signature   : (none)
Packager    : Chef Supermarket Team <supermarket@getchef.com>
URL         : https://supermarket.getchef.com
Summary     : The full stack of supermarket
Description :
The full stack of supermarket
```

### CentOS 7
```
[vagrant@slave-centos-70 ~]$ cat /etc/redhat-release
CentOS Linux release 7.0.1406 (Core)
[vagrant@slave-centos-70 ~]$ rpm --version
RPM version 4.11.1
[vagrant@slave-centos-70 ~]$ sudo rpm -Uvh supermarket-1.10.1~alpha.0-1.el5.x86_64.rpm
Preparing...                          ################################# [100%]
Updating / installing...
   1:supermarket-1.10.1~alpha.0-1.el5 ################################# [100%]

Thank you for installing Supermarket!

The next step in the process is to run:

supermarket-ctl reconfigure

[vagrant@slave-centos-70 ~]$ rpm -qi supermarket
Name        : supermarket
Version     : 1.10.1~alpha.0
Release     : 1.el5
Architecture: x86_64
Install Date: Sat 06 Jun 2015 08:44:10 PM UTC
Group       : default
Size        : 1141070737
License     : unknown
Signature   : (none)
Source RPM  : supermarket-1.10.1~alpha.0-1.el5.src.rpm
Build Date  : Mon 18 May 2015 08:04:01 PM UTC
Build Host  : ip-10-194-12-70.us-west-2.compute.internal
Relocations : /
Packager    : Chef Supermarket Team <supermarket@getchef.com>
Vendor      : Omnibus <omnibus@getchef.com>
URL         : https://supermarket.getchef.com
Summary     : The full stack of supermarket
Description :
The full stack of supermarket
```

**Comment by irvingpop** [comment]
hey @schisamo totally agree with you.  I've worked up a fix here: https://github.com/chef/chef/pull/3503  and would appreciate if you could help me validate it

**Comment by lamont-granquist** [comment]
It looks like #3503 fixed the bleeding here and #3671 is a better tracking ticket for the problem of inverting the match, so closing this one.
